server/src/main/java/org/apache/uniffle/server/ShuffleServer.java

```diff
@@ -259,6 +274,79 @@ private void blockUntilShutdown() throws InterruptedException {
     server.blockUntilShutdown();
   }
 
+  public ServerStatus getServerStatus() {
+    return serverStatus;
+  }
+
+  public synchronized void decommission() {
+    if (isDecommissioning()) {
+      LOG.info("Shuffle Server is decommissioning. Nothing needs to be done.");
+      return;
+    }
+    if (!ServerStatus.ACTIVE.equals(serverStatus)) {
+      throw new InvalidRequestException(
+          "Shuffle Server is processing other procedures, current status:" + serverStatus);
+    }
+    serverStatus = ServerStatus.DECOMMISSIONING;
+    LOG.info("Shuffle Server is decommissioning.");
+    if (executorService == null) {
+      executorService = Executors.newSingleThreadExecutor(
+          ThreadUtils.getThreadFactory("shuffle-server-decommission-%d"));
+    }
+    decommissionFuture = executorService.submit(this::waitDecommissionFinish);
+  }
+
+  private void waitDecommissionFinish() {
+    long checkInterval = shuffleServerConf.get(SERVER_DECOMMISSION_CHECK_INTERVAL);
+    boolean shutdownAfterDecommission = shuffleServerConf.get(SERVER_DECOMMISSION_SHUTDOWN);
+    int remainApplicationNum;
+    while (isDecommissioning()) {
+      remainApplicationNum = shuffleTaskManager.getAppIds().size();
+      if (remainApplicationNum == 0) {
+        serverStatus = ServerStatus.DECOMMISSIONED;
+        LOG.info("All applications finished. Current status is " + serverStatus);
+        if (shutdownAfterDecommission) {
+          LOG.info("Exiting...");
+          try {
+            stopServer();
+          } catch (Exception e) {
+            ExitUtils.terminate(1, "Stop server failed!", e, LOG);
+          }
+        }
+        break;
+      }
+      LOG.info("Shuffle server is decommissioning, remaining {} applications not finished.", remainApplicationNum);
+      try {
+        Thread.sleep(checkInterval);
+      } catch (InterruptedException e) {
+        LOG.warn("Ignore the InterruptedException which should be caused by internal kill.");
+      }
+    }
+    remainApplicationNum = shuffleTaskManager.getAppIds().size();
+    if (remainApplicationNum > 0) {
+      LOG.info("Decommission exiting, remaining {} applications not finished.",
+          remainApplicationNum);
+    }
+  }
+
+  public synchronized void cancelDecommission() {
+    if (!isDecommissioning()) {
+      LOG.info("Shuffle server is not decommissioning. Nothing needs to be done.");
+      return;
+    }
+    if (ServerStatus.DECOMMISSIONED.equals(serverStatus)) {
+      serverStatus = ServerStatus.ACTIVE;
+      return;
+    }
+    serverStatus = ServerStatus.ACTIVE;
+    if (decommissionFuture.cancel(true)) {
+      LOG.info("Decommission canceled.");
+    } else {
+      LOG.warn("Failed to cancel decommission.");
+    }
```

Comment:
The result of `decommissionFuture.cancel()` doesn't matter now. Since "cancel" always brings a server back to ACTIVE. Maybe we should call `executorService.shutdownNow()` here, @advancedxy WDYT?

```suggestion
    if (decommissionFuture != null) {
      decommissionFuture.cancel(true);
    }
```